src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   more pending promises, triggering more callbacks, etc. This behavior is the
+   *completion loop*. Lwt has some machinery to avoid stack overflow and other
+   unfortunate situations during this loop.
+
+   This chaining of promise completions through callbacks can be seen as a kind
+   of promise dependency graph, in which the nodes are pending promises, and the
+   edges are callbacks. During the completion loop, Lwt starts at some initial
+   promise that is getting completed by the user, and completes a bunch of
+   dependent promises.
+
+   Some of these dependencies are explicit to Lwt, e.g. the callbacks registered
+   by [Lwt.bind]. Others are not visible to Lwt, because the user can always
+   register a callback using a function like [Lwt.on_success], and use that
+   callback to complete another initial promise. All the explicit dependencies
+   are created by Lwt's own sequential and concurrent composition functions
+   (so, [Lwt.bind], [Lwt.join], etc). Whether dependencies are explicit or not
+   is relevant only to cancelation.
+
+
+   5. Cancelation
+
+   As described above, ordinary promise completion procedes from an initial
+   promise, forward along callbacks through the dependency graph. Since it
+   starts from an initial promise, it can only be triggered using a resolver.
+
+   Cancelation is a sort of dual to ordinary completion. Instead of starting at
+   an initial promise/resolver, cancelation starts at *any* promise. It then
+   goes *backwards* through the explicit dependency graph, looking for
+   cancelable initial promises to cancel -- those that were created by
+   [Lwt.task]. After finding them, cancelation completes them normally with
+   [Failed Lwt.Canceled], causing an ordinary promise completion process.
+
+   To summarize, cancelation is a way to trigger an *ordinary* completion of
+   promises created with [Lwt.task], by first searching for them in the promise
+   dependency graph (which is assembled by [Lwt.bind], [Lwt.join], etc).
+
+   This backwards search is triggered only by [Lwt.cancel]. It is also possible
+   for the user to cancel a promise directly by failing it with [Lwt.Canceled],
+   but in all cases where the user can do so, the search would be redundant
+   anyway -- the user has only two ways of directly failing a promise with
+   [Lwt.Canceled] (or any exception, for that matter):
+
+   - The user can create an initial promise, then fail it through its resolver.
+     The search is redundant because it would find only the same initial promise
+     to cancel.
+   - The user can create a trivial promise by calling [Lwt.fail Lwt.Canceled].
+     The search is again redundant; in this case it would find nothing to
+     cancel.
+
+   Note that there is a quirk: only promises created by [Lwt.task] are
+   susceptible to being canceled by [Lwt.cancel], but the user can manually
+   cancel initial promises created by both [Lwt.task] and [Lwt.wait].
```

Comment:
Indeed, I've added a comment to `bind`: https://github.com/ocsigen/lwt/blob/d1abbf586f3eb82a9b0275fc204fdf85d728249a/src/core/lwt.ml#L1732-L1738